drivers: nrf_wifi: Restructure to carve out a new nrf70 bus lib

bama-nordic opened the pull request:
The nrf wifi drivers include code related to qspi/spi bus. This work moves bus related files to a new nrf70 bus library which is now independent of the core driver and facilitlates building applications directly using the bus APIs, agnostic to enabling wifi configs.

krish2718 commented on modules/nrf_wifi/bus/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+  zephyr_library_sources_ifdef(CONFIG_NRF70_ON_SPI
+    spi_if.c
+  )
+  zephyr_library_sources(device.c)
```
this can be combined with `rpu_hw_if.c`

krish2718 commented on modules/nrf_wifi/bus/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+if (CONFIG_NRF70_BUSLIB)
+
+  zephyr_library_named(nrf70-buslib)
+
```
no need for empty lines

krish2718 commented on modules/nrf_wifi/bus/Kconfig:
```diff
@@ -0,0 +1,57 @@
+        bool
+        default y if DT_HAS_NORDIC_NRF7001_SPI_ENABLED || DT_HAS_NORDIC_NRF7001_QSPI_ENABLED
+
+config WIFI_NRF7000
```
these are only used in native driver

krish2718 commented on modules/nrf_wifi/bus/Kconfig:
```diff
@@ -0,0 +1,57 @@
+	def_bool $(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF70_SPI))
+	select SPI
+
+#config NRF70_QSPI_LOW_POWER
+#	bool "low power mode in QSPI"
+#	default y if NRF_WIFI_LOW_POWER
```
local change?

bama-nordic replied:
moved to NRF70 driver Kconfig

krish2718 commented on modules/nrf_wifi/bus/rpu_hw_if.c:
```diff
@@ -483,10 +483,12 @@ int rpu_enable(void)
 		goto rpu_pwroff;
 	}
 
+#if 0
```
local change?

bama-nordic replied:
We cannot have `rpu_validate_comms()` anymore since it will not work unless firmware is loaded and for a independent bus lib, this wont work. I think its best we move this function as we will have bustest now, plan to do in the next commit

bama-nordic replied:
OK, I have put this under #ifdef CONFIG_WIFI_NRF70 - that should take care of the issue while not changing functionality for the core wifi driver.

krish2718 commented on modules/nrf_wifi/bus/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+  )
+  zephyr_library_sources(device.c)
+
+  set_target_properties(nrf70-buslib PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${CMAKE_CURRENT_SOURCE_DIR})
```
`zephyr_library_include_directories` should take care of this automatically.

krish2718 reviewed (commented):
remove `ficr_prog.*` should be a separate commit as it's unrelated to buslib.

bama-nordic committed:
drivers: nrf_wifi: Restructure to carve out a new nrf70 bus lib

The nrf wifi drivers include code related to qspi/spi bus. This work
moves bus related files to a new nrf70 bus library which is now
independent of the core driver and facilitlates building applications
directly using the bus APIs, agnostic to enabling wifi configs.

Signed-off-by: Bansidhar P.M <bansidhar.mangalwedhekar@nordicsemi.no>

bama-nordic committed:
tests: boards: nrf70: Add ztests for bus library

Add ztests for nrf70 bus lib to help validate functional integrity
of the QPSI/SPI interface to nRF70 device via appropriate reads/writes
from host processor to relevant memory blocks of nrf70 device.
Note that this will NOT serve as a memory test for nrf70 device,
rather just a functional verification of the wiring between the host
processor and the nrf70 device.

Signed-off-by: Bansidhar P.M <bansidhar.mangalwedhekar@nordicsemi.no>

krish2718 commented on modules/nrf_wifi/bus/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: BSD-3-Clause
```
this should also be apache-2.0 for scan code to pass.

krish2718 commented on modules/nrf_wifi/bus/Kconfig:
```diff
@@ -0,0 +1,38 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+DT_COMPAT_NORDIC_NRF70_QSPI := nordic,nrf7002-qspi
```
This won't work with nRF7000/nRF7001, need to add entries for both.

bama-nordic replied:
Added both now

jukkar commented on modules/nrf_wifi/bus/qspi_if.c:
```diff
@@ -195,7 +190,7 @@ struct qspi_nor_data {
 	/* The semaphore to control driver init/uninit. */
 	struct k_sem count;
-#else /* CONFIG_MULTITHREADING */
+#else  /* CONFIG_MULTITHREADING */
```
Could you remove these unnecessary changes. I see it here and below in multiple places.

jukkar commented on modules/nrf_wifi/bus/qspi_if.c:
```diff
@@ -688,7 +681,7 @@ static inline void qspi_fill_init_struct(nrfx_qspi_config_t *initstruct)
 #endif /* CONFIG_PINCTRL */
 	/* Configure Protocol interface */
-	initstruct->prot_if.addrmode = NRF_QSPI_ADDRMODE_24BIT;
+		initstruct->prot_if.addrmode = NRF_QSPI_ADDRMODE_24BIT;
```
Indent is now wrong

jukkar commented on modules/nrf_wifi/bus/qspi_if.c:
```diff
@@ -1000,28 +993,23 @@ static int qspi_nor_configure(const struct device *dev)
 #ifdef CONFIG_PINCTRL
-	int ret = pinctrl_apply_state(QSPI_IF_BUS_PINCTRL_DT_DEV_CONFIG_GET,
-				      PINCTRL_STATE_DEFAULT);
+	int ret = pinctrl_apply_state(QSPI_IF_BUS_PINCTRL_DT_DEV_CONFIG_GET, PINCTRL_STATE_DEFAULT);
```
Did you run this through clang-format? It does these unnecessary changes, please remove them.

jukkar reviewed (changes requested):
Could you remove all cosmetic and unnecessary changes, they make it hard to review this.

fabiobaltieri commented on modules/nrf_wifi/bus/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: BSD-3-Clause
```
Hi, this has not been fixed and is still BSD-3-Clause upstream, @krish2718 would you be able to send a PR to fix it?

https://github.com/zephyrproject-rtos/zephyr/blob/c7a19da8ca1abefa8eac41e704e7c678af0282a9/modules/nrf_wifi/bus/CMakeLists.txt#L4

cc @kartben

krish2718 replied:
FYI https://github.com/zephyrproject-rtos/zephyr/pull/86748
